neuralpp/symbolic/quantifier_free_normalizer.py

```diff
@@ -0,0 +1,72 @@
+from typing import Optional, Deque
+from collections import deque
+
+from .expression import Expression, Constant, FunctionApplication
+from .z3_expression import Z3SolverExpression
+from .constants import basic_true, basic_false, if_then_else
+from .context_simplifier import ContextSimplifier
+from .parameters import sympy_evaluate
+from .normalizer import Normalizer
+
+
+def is_non_constant_formula(expression: Expression) -> bool:
+    return expression.type == bool and not isinstance(expression, Constant)  # should ignore `True` and `False`
+
+
+def _bfs_first_non_constant_formula_of_expression(expression: Expression, queue: Deque) -> Optional[Expression]:
+    queue.append(expression)
+    while queue:
+        current_expression = queue.popleft()
+        if is_non_constant_formula(current_expression):
+            return current_expression
+        match current_expression:
+            case FunctionApplication(arguments=arguments):
+                # This includes if clause in a conditional expression, we need to split on that
+                # to simplify its subexpressions.
+                # E.g. if x > 3 then (if x > 2 then .. else ..) else ..
+                # if we do not split on x > 3, then the simplification x > 2 -> True cannot be made
+                queue.extend(arguments)
+            case _:
+                pass  # non_constant_formulaic expression has no subexpressions
+    return None
+
+
+def first_non_constant_formula_of_expression(expression: Expression) -> Optional[Expression]:
+    """ returns the first non_constant_formula occurring in a given expression in a breadth-first search. """
+    return _bfs_first_non_constant_formula_of_expression(expression, deque())
+
+
+class QuantifierFreeNormalizer(Normalizer):
```

Comment:
This is reverted to an old version of `Normalizer` before we introduce and consider the quantifier case.